### PR TITLE
Add Nonparametric PDA and gate out non valid measurements

### PR DESCRIPTION
**Pull request by sdhiscocks**
This adds the ability to use nonparametric PDA for clutter spatial density, where it is estimated based on assumption that measurements within a defined validation region represent the clutter level.

This change also gates out non valid measurements by default, similar to the behaviour in the `DistanceHypothesier`, and per description in original PDA paper.

### CODE REVIEW

File: stonesoup/hypothesiser/probability.py [hunk 1]
```diff
@@ -118,6 +137,7 @@ def hypothesise(self, track, detections, timestamp, **kwargs):
 
         # True detection hypotheses
         for detection in detections:
+            valid_measurement = False
```
Comment (sglvladi) [inline stonesoup/hypothesiser/probability.py:140]: Maybe move this above line 154 with some comments about gating?

File: stonesoup/hypothesiser/probability.py [hunk 2]
```diff
@@ -126,18 +146,57 @@ def hypothesise(self, track, detections, timestamp, **kwargs):
...
+    @lru_cache()
+    def _validation_region_volume(cls, prob_gate, meas_pred):
+        n = meas_pred.ndim
+        gate_threshold = cls._gate_threshold(prob_gate, n)
+        c_z = np.pi**(n/2) / gamma(n/2 + 1)
+        return c_z * gate_threshold**(n/2) * np.sqrt(det(meas_pred.covar))
+
+    @staticmethod
+    @lru_cache()
+    def _inv_cov(meas_pred):
```
Comment (sglvladi) [inline stonesoup/hypothesiser/probability.py:196]: Not too keen on single-line functions, but I'm guessing this is done to take advantage of caching?

File: stonesoup/hypothesiser/probability.py [hunk 3]
```diff
@@ -126,18 +146,57 @@ def hypothesise(self, track, detections, timestamp, **kwargs):
...
+                valid_measurement = True
+
+            if self.include_all or valid_measurement:
+                probability = pdf * self.prob_detect
+                if self.clutter_spatial_density is None:
+                    # Note: will divide by validated measurements count later...
+                    probability *= self._validation_region_volume(
+                        self.prob_gate, measurement_prediction)
+                else:
+                    probability /= self.clutter_spatial_density
```
Comment (sglvladi) [inline stonesoup/hypothesiser/probability.py:165]:
```suggestion
                if self.clutter_spatial_density is not None:
                    probability /= self.clutter_spatial_density
```

Might be more readable if the calculation for the case ``self.clutter_spatial_density is None`` is all done in one place on lines 175-177, given that we are iterating over the hypotheses and making the check anyway. The comment is helpful, but I think keeping the "equations" together is more helpful.

**Comment by sglvladi** [comment]
Looks good overall 👍 Added some minor notes on making the code more readable.

Note on earlier discussion: I don't think we can avoid gating when doing the non-parametric version. We could avoid it for the parametric one, but then it would be confusing to say we do gating in one case but not the other.

File: stonesoup/hypothesiser/probability.py (hunk 2 above)
Comment (sdhiscocks) [reply stonesoup/hypothesiser/probability.py:196]: Yes. However, there are other modifications with Mahalanobis measure that include built in caching. I may rebase of main and then can take advantage of that and remove the need for this function.

**Review by sglvladi — APPROVED** [review]

**Review by jswright-dstl — APPROVED** [review]

**Pull request merged**